contracts/src/registry/EnhancedAccessControl.sol

```diff
@@ -0,0 +1,302 @@
+    event EnhancedAccessControlRolesRevoked(bytes32 resource, uint256 role, address account, address sender);
+    event EnhancedAccessControlAllRolesRevoked(bytes32 resource, address account, address sender);
+    event EnhancedAccessControlRolesLocked(bytes32 resource, uint256 roleBitmap);
+
+    /** 
+     * @dev admin role that controls a given role. 
+     * RoleId -> AdminRoleId
+     */
+    mapping(uint256 role => uint256 adminRole) private _adminRoles;
+
+    /**
+     * @dev user roles within a resource stored as a bitmap.
+     * Resource -> User -> RoleBitmap
+     */
+    mapping(bytes32 resource => mapping(address account => uint256 roleBitmap)) private _roles;
+
+    /**
+     * @dev locked roles within a resource stored as a bitmap.
+     * Resource -> LockedRoleBitmap
+     */
+    mapping(bytes32 resource => uint256 lockedRoleBitmap) private _lockedRoles;
+
+    /**
+     * @dev The root resource.
+     */
+    bytes32 public constant ROOT_RESOURCE = bytes32(0);
+
+    /**
+     * @dev The default admin role.
+     */
+    uint256 public constant DEFAULT_ADMIN_ROLE = 1;
+
+    /**
+     * @dev Modifier that checks that sender has the admin role for the given role. 
+     * If the sender does not have the admin role, it checks that the sender has the DEFAULT_ADMIN_ROLE.
+     */
+    modifier canGrantRole(bytes32 resource, uint256 role) {
+        _checkCanGrantRole(resource, role, _msgSender());
+        _;
+    }
+
+    /**
+     * @dev Modifier that checks that sender has a specific role within the given resource. 
+     * Reverts with an {AccessControlUnauthorizedAccount} error including the required role.
+     */
+    modifier onlyRole(bytes32 resource, uint256 role) {
+        _checkRole(resource, role, _msgSender());
+        _;
+    }
+
+    /**
+     * @dev Modifier that checks that sender has a specific role within the root resource. 
+     * Reverts with an {AccessControlUnauthorizedAccount} error including the required role.
+     */
+    modifier onlyRootRole(uint256 role) {
+        _checkRole(ROOT_RESOURCE, role, _msgSender());
+        _;
+    }
+
+    /**
+     * @dev Constructor.
+     *
+     * @param initialAdmin The address to grant the DEFAULT_ADMIN_ROLE to.
+     */
+    constructor(address initialAdmin) {
+        _grantRoles(ROOT_RESOURCE, DEFAULT_ADMIN_ROLE, initialAdmin);
+    }
+
+    /**
+     * @dev See {IERC165-supportsInterface}.
+     */
+    function supportsInterface(bytes4 interfaceId) public view virtual override returns (bool) {
+        return interfaceId == type(EnhancedAccessControl).interfaceId || super.supportsInterface(interfaceId);
+    }
+
+    /**
+     * @dev Returns `true` if `account` has been granted all roles in the root resource.
+     */
+    function hasRootRoles(uint256 rolesBitmap, address account) public view virtual returns (bool) {
+        return (_roles[ROOT_RESOURCE][account] & rolesBitmap) != 0;
```

Comment:
```suggestion
        return (_roles[ROOT_RESOURCE][account] & rolesBitmap) == rolesBitmap;
```

As it is it checks if the user has *any* of the provided roles, but I think we want to check if the user has *all* of the provided roles.